### ISSUE TITLE
Support for `wavedecn` and `waverecn` [WAS: Hi-Lo and Lo-Hi filters swapped in 2D transform]

### ISSUE THREAD

**Issue by Obi-Wan** [v0lta/PyTorch-Wavelet-Toolbox#114]
### Description
The two mixed-scale coefficients hi-lo and lo-hi are swapped in the 2D discrete wavelet transform. This is not a big deal if we're just using `ptwt` (as it will always be self-consistent), but it results in incorrect reconstructions when mixing `ptwt` and `pywt` (unless we do some argument "gymnastics" when translating from one to the other).

https://github.com/v0lta/PyTorch-Wavelet-Toolbox/blob/d62c62c4e8ee349a9c0bb4576435b811a7fad20a/src/ptwt/conv_transform_2.py#L49-L50

### Proposed solution
I propose to fix this issue by simply correcting the order of the filters at those two lines. This could result in backwards incompatibility, but only in the case where the coefficients were to be saved and restored. For any on-the-fly analysis, it should be pretty transparent. I would suggest issuing a warning at the time of the next release.

### Possible further improvement
It would be great if the return value of `wavedec2` and the first argument to `waverec2` could be of type `WaveletCoeffNd` instead of `WaveletCoeff2d`. This would ensure complete consistency (and easier compatibility) against `pywt`.

**Comment by felixblanke** [comment]
Dear @Obi-Wan, thanks for raising the issue!

I just checked with this toy example and for me the order of dimensions seems consistent to the `pywt` implementation.
```python
import numpy as np
import pywt, ptwt, scipy, torch
from torch.testing import assert_close

# example image, move to dims [C, H, W]
face = scipy.datasets.face().transpose([2, 0, 1]).astype(np.float64)
pytorch_face = torch.tensor(face, dtype=torch.float64)

ptwt_A, (ptwt_H, ptwt_V, ptwt_D) = ptwt.wavedec2(
    pytorch_face, "db4", level=1, mode="reflect"
)

pywt_A, (pywt_H, pywt_V, pywt_D) = pywt.wavedec2(
    face, "db4", level=1, mode="reflect",
)

# no asserts triggered:
assert_close(ptwt_A, torch.from_numpy(pywt_A))
assert_close(ptwt_H, torch.from_numpy(pywt_H))
assert_close(ptwt_V, torch.from_numpy(pywt_V))
assert_close(ptwt_D, torch.from_numpy(pywt_D))

# the following would raise an assert
# assert_close(ptwt_V, torch.from_numpy(pywt_H))
# assert_close(ptwt_H, torch.from_numpy(pywt_V))
```
We also check against the direct `pywt` equivalent in multiple unit test cases, so I think we don't have a general inconsistency to `pywt`. However, I agree that the "order" at the code you linked to looks at least unintuitive. I will look into that and come back to you.

Regarding the suggested change of return types: I agree that always returning dicts would be more consistent. But the deviation in the 2d case directly mirrors the design of pywt, [see their API docs of e.g. `pywt.wavedec2`](https://pywavelets.readthedocs.io/en/latest/ref/2d-dwt-and-idwt.html#pywt.wavedec2)

**Comment by Obi-Wan** [comment]
Thank you for your quick reply. I must admit that at first I was a bit puzzled by your result. I had a working example that clearly required me to swap horizontal and vertical. However, I just realized that I was using `wavedecn` and `waverecn` (which BTW always work with the dictionary form).

I'm sure the mess-up might come from the nomenclature `da` vs `ad` from `pywt`'s `waverecn` and the corresponding `hl` vs `lh` in `ptwt`'s `waverec2`.

I suppose I should then change the title into: "Are you planning on supporting `wavedecn` and `waverecn`?"
On the same note, I would also be interested in `swtn` and `iswtn`. Any plans?

**Comment by felixblanke** [comment]
**Regarding n-dim transforms:**
For the convolutional backend, we use the pytorch convolution operators under the hood. Right now, these are only implemented up to 3d and have no support for arbitrary dimensions. So until the respective kernels are implemented upstream, I would say the answer right now is a no.

What is supported at the moment is the fully separable convolutional transform (i.e. 1d convs along each axis). This is implemented in:
https://github.com/v0lta/PyTorch-Wavelet-Toolbox/blob/d62c62c4e8ee349a9c0bb4576435b811a7fad20a/src/ptwt/separable_conv_transform.py#L352
(which is only a private func for consistency with other parts of the code base). Theoretically, we could also extend the sparse matrix backend for arbitrary dimensions, if there is demand for it.

**Regarding the nomenclature:**
I looked a bit through pywavelets docs and found two different assignments of LH/HL to H/V:
1. in the [dwtn docs](https://pywavelets.readthedocs.io/en/latest/ref/nd-dwt-and-idwt.html#pywt.dwtn): `V(LH)` and `H(HL)`
2. in the [dwt2 docs](https://pywavelets.readthedocs.io/en/latest/ref/2d-dwt-and-idwt.html#single-level-dwt2), there they use `cH(LH)` and `cV(HL)`.

This is somewhat confusing but I think the first is the correct naming (this is also calculated in the code you linked to above). I checked and we don't mention this nomenclature in the docs (but the assignment to the axes is correctly stated in the relevant docs page). I will look into refactoring internal naming to make it consistent

**Comment by Obi-Wan** [comment]
Ok, thank you very much for the answers. I am implementing what I need, starting from the functions you implemented, and picking as much as possible from the existing supporting functions. This package's code (especially the `swt`) is very well written, and exhibits very good performance.
